samples/coop_synchronization/src/main.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. (AMD)
+ * Copyright (c) 2023 Alp Sayin <alpsayin@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+/*
+ * The coop synchronization demo has two threads that utilize semaphores and k_yield
+ * to take turns printing a greeting message with a somewhat uncontrolled rate.
+ * The rate is controlled via a busy waiting loop. The advantage of this demo is that
+ * it'll work even if there are no timing utilities provided by the architecture.
+ */
+
+#define PIN_THREADS (IS_ENABLED(CONFIG_SMP)		  \
+		     && IS_ENABLED(CONFIG_SCHED_CPU_MASK) \
+		     && (CONFIG_MP_NUM_CPUS > 1))
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+/*
+ * @param my_name      thread identification string
+ * @param my_sem       thread's own semaphore
+ * @param other_sem    other thread's semaphore
+ */
+void helloLoop(const char *my_name, struct k_sem *my_sem,
+	struct k_sem *other_sem, int incrementer)
+{
+	const char *tname;
+	uint8_t cpu;
+	struct k_thread *current_thread;
+	int task_local_counter = 0;
+
+	while (1) {
+		/* take my semaphore */
+		k_sem_take(my_sem, K_FOREVER);
+
+		int key1 = irq_lock();
+		{
+			current_thread = k_current_get();
+			tname = k_thread_name_get(current_thread);
+#if CONFIG_SMP
+			cpu = arch_curr_cpu()->id;
+#else
+			cpu = 0;
+#endif
```

Comment:
Nitpick: `arch_curr_cpu()` works fine when SMP=n and there is only one CPU, and its ID is indeed zero.  This doesn't seem like a performance path, so probably best to remove the #ifery.

Reply:
Trying to build for `qemu_x86` with sdk16 gives me below error:
```/home/asayin/zephyr_sdks/microblaze_devel/zephyr/samples/coop_synchronization/src/main.c: In function 'helloLoop':
/home/asayin/zephyr_sdks/microblaze_devel/zephyr/samples/coop_synchronization/src/main.c:48:23: warning: implicit declaration of function 'arch_curr_cpu'; did you mean 'arch_num_cpus'? [-Wimplicit-function-declaration]
   48 |                 cpu = arch_curr_cpu()->id;
      |                       ^~~~~~~~~~~~~
      |                       arch_num_cpus
/home/asayin/zephyr_sdks/microblaze_devel/zephyr/samples/coop_synchronization/src/main.c:48:38: error: invalid type argument of '->' (have 'int')
   48 |                 cpu = arch_curr_cpu()->id;
      |                                      ^~
```
It seems x86 arch only defines `arch_curr_cpu()` for 64-bit (if CONFIG_X86_64).
From what I see `arch_curr_cpu()` is defined only for ARC, ARM, X86_64, RISV and XTENSA.
Also from what I see,  `arch_curr_cpu()` is defined only if `CONFIG_SMP` in `include/zephyr/sys/arch_interface.h`.
I tried to come up with a more sensible barrier but the whole situation seems messy...
```
#if defined(CONFIG_ARC) || defined(CONFIG_ARM) || defined(CONFIG_X86_64) \
	|| defined(CONFIG_RISCV) || defined(CONFIG_XTENSA)
		cpu = arch_curr_cpu()->id;
#endif
```
Let me know what you think please.

p.s. This also affects the original `synchronization` sample.